Ticket 2: Develop backend API endpoints for user registration and storing user data

chenyou-H opened the pull request:
Implement all required features and an integration test for API.

I found may be some setting problems between Mongoose and the jest test environment.
!! In order to run the user-signup.test.ts, I have to make a modification in the node_modules.
//Open the file in the following link /node_modules/whatwg-url/lib/encoding.js
//and add--> const { TextEncoder, TextDecoder } = require('util') under the 'use Strict'.

Otherwise, when you try to connect to the database in the test suites, you may the error like:   
● Test suite failed to run
    ReferenceError: TextEncoder is not defined.

TextEncoder is a node built-in library, we are not supposed to have this error. We have this error because the jest test suite doesn't use node.js as its testing environment. I am not sure how to fix this problem by changing the setting. Maybe we should create a ticket for this

NobodyLand committed:
Implement user schema

NobodyLand committed:
create the api file

NobodyLand committed:
implememt create user in database but database not working well

NobodyLand committed:
Implement database store

NobodyLand committed:
Check if email existed before creating new user

NobodyLand committed:
Create test files

NobodyLand committed:
install node-mocks-http in order to mock api request and response in test

songhany committed:
comments

NobodyLand committed:
implement validation function test

NobodyLand committed:
Merge https://github.com/VijayIyer/skool-clone into user-api

NobodyLand committed:
Unable to implement test for next js api

NobodyLand committed:
Implement test case for sign up api

Commit:
Revert "comments"

This reverts commit e6674b2bdc1452ebf58ae9275d0ed02a5ff176f3.

PHANTOMS0308 commented on models/User.js:
```diff
@@ -0,0 +1,28 @@
+  timestamps: true, // this will add createdAt and updatedAt timestamps
+});
+
+export default mongoose.models.User || mongoose.model('User', UserSchema)
```
Love this little detail to prevent the potential error of **OverwriteModelError**

PHANTOMS0308 commented on lib/userApi.ts:
```diff
@@ -0,0 +1,53 @@
+import User from '../models/User';
+
+//User CRUD function below can only be invoke when server connects with database, such as code like  "await dbConnect();" from lib/dbConnect.ts is invoked
+export async function createNewUser(
```
`createUser` might be self-explanatory, since when you create user, that must be a new user 🤣 

Also I believe having this function taking an object will be better for developer experience.
For instance, with TS you can define an interface or type of UserSignupInput and be used both in `createUser()` and `validateUserSignupInput()` inside your `./validation.ts`

```typescript
interface UserSignupInput {
  lastName: string;
  firstName: string;
  email: string;
  password: string;
}
```

PHANTOMS0308 commented on models/User.js:
```diff
@@ -0,0 +1,28 @@
+import mongoose from 'mongoose'
+
+const UserSchema = new mongoose.Schema({
+  firstname: {
```
This might be picky, but camelCase feels better for the field, like `firstName` since we usually write "first name" as two letters in usual writing

PHANTOMS0308 commented on models/User.js:
```diff
@@ -0,0 +1,28 @@
+
+}, {
+  timestamps: true, // this will add createdAt and updatedAt timestamps
+});
```
There could be one business logic, that comes very in handy. If you look at the Skool website, in most scenario user name are shown as a full name ( firstName, lastName )

For example, `{ firstName: 'Shengfeng', lastName: 'Wu' }` is shown as 'Shengfeng Wu'. Therefore, we may have this business logic binding to the schema, using `virtuals`

```javascript
const userSchema = new Schema({
/// ...schemaFields
}, {
  virtuals: {
    fullName: {
      get() {
        return this.firstName + ' ' + this.lastName;
      }
    }
  }
});

// suppose user is an instance of User Model
console.log(user.fullName)
```

You can think of this virtuals is exactly like the get() and set() function that we define in Javascript Class, and the `fullName` will only be stored inside the userSchema not in the actual database, thus named virtuals

PHANTOMS0308 commented on models/User.js:
```diff
@@ -0,0 +1,28 @@
+import mongoose from 'mongoose'
+
+const UserSchema = new mongoose.Schema({
```
Usually only constructor are capitalized camel case, so it is better to be `userSchema`, because this is an instance of `mongoose.Schema` not an instructor.

PHANTOMS0308 commented on lib/validation.ts:
```diff
@@ -0,0 +1,58 @@
+  return validator.isEmail(email);
+}
+
+export function validatePassword(password: string) {
```
The validation logic should be consistent with front end one, password should be in the range of [5, 72] inclusive.

PHANTOMS0308 commented on lib/validation.ts:
```diff
@@ -0,0 +1,58 @@
+  return password.length >= 5;
+}
+
+export function validateName(name: string) {
```
Same, both the first name and the last name should be <= 20 characters

PHANTOMS0308 commented on lib/validation.ts:
```diff
@@ -0,0 +1,58 @@
+  password: string;
+}) {
+  if (!validateName(first_name)) {
+    return {
```
This error object returned to the controllers is better to be centralized into one place, for better consistency. For example we can define class like ModelError that allow you to pass in some parameter. The purpose is for better code reusability and consistency throughout the whole backend. But Alex you don't need to fix this, it could be integrated later on

PHANTOMS0308 commented on lib/validation.ts:
```diff
@@ -0,0 +1,58 @@
+import validator from 'validator';
+
+export function validateUserSignUpInput({
+  first_name,
```
Also camel case should be used, for better code consistency, we should stick with one naming convention

PHANTOMS0308 commented on models/User.js:
```diff
@@ -0,0 +1,28 @@
+    type: String,
+    required: [true, 'Please provide a first name for this user']
+  },
+  lastname: {
```
If you observe closely on the Skool UI, almost every user last name and first name are capitalized. However, neither did Zhiyuan's signup form nor your signup API, purify the user's last name and first name. Since this is more of a business logic, it could be added to the schema. Below is an example from the Mongoose Doc

```javascript
function capitalize (name) {
  if (typeof name !== 'string') name = '';
  return name.charAt(0).toUpperCase() + name.substring(1);
}

// defining within the schema
const userSchema = new Schema({ 
  // ...other fields
  lastName: { type: String, set: capitalize }
});
```

PHANTOMS0308 commented on pages/api/comments.ts:
Maybe this was a merge mistake, I will just ignore it anyway

PHANTOMS0308 commented on pages/api/user-signup.ts:
```diff
@@ -0,0 +1,64 @@
+  switch (req.method) {
+    case 'POST':
+      try {
+        const user = JSON.parse(req.body);
```
> req.body - An object containing the body parsed by content-type, or null if no body was sent

This paragraph is from the[ NextJS doc](https://nextjs.org/docs/pages/building-your-application/routing/api-routes), so there is no need to use `JSON.parse()` here

PHANTOMS0308 commented on pages/api/user-signup.ts:
```diff
@@ -0,0 +1,64 @@
+        if (!validationResult.success) {
+          return res
+            .status(400)
+            .json({ success: false, message: validationResult.message });
```
Also here, should we have a centralized logic for all error object sending back to the front end developer? That should be something to consider sooner or later.

PHANTOMS0308 commented on pages/api/user-signup.ts:
```diff
@@ -0,0 +1,64 @@
+
+        const hashedPassword = await generateHashPassword(user.password);
+        const dbRes = await createNewUser(
+          user.first_name,
```
So if this user is from `req.body` then you are assuming that front end developer will be using snake_case. So again this causes naming inconsistency, try to make everything camelCase

PHANTOMS0308 commented on lib/hashPassword.ts:
```diff
@@ -0,0 +1,16 @@
+//create this function for login api
+export async function compareHashPasswordWith(email: String, password: String) {
+  const user = await findUserByEmail(email);
+  const savedPassword = user[0].password;
```
Since your method is named `findUserByEmail` and also `email` is a unique throughout the collection, then developer expect this function to return an instance of user, not an array of user instances, though that array will be length of 1. Maybe it is better to redesign your `findUserByEmail()` to make sure it either return an Mongoose object or throw error. You may try `Model.findOne()` instead of `Model.find()`

PHANTOMS0308 commented on __test__/pages/user-signup.test.ts:
For this test, I will try to recreate that Error on my end, and also try to narrow down what are the possible causes that Lynnda did not encounter the same Error.

PHANTOMS0308 replied:
Also this folder should not be named `/pages` but `/api`

PHANTOMS0308 commented on pages/api/user-signup.ts:
```diff
@@ -0,0 +1,64 @@
+    await dbConnect();
+  } catch (e) {
+    console.log('fail to connect', e);
+    // console.error(e);
```
Shouldn't here have a `return res.status(500).json({ success: false, message: 'Server Side Error...'})` 
Cuz right now if there was database error, then your handler will hang there

PHANTOMS0308 commented on pages/api/user-signup.ts:
```diff
@@ -0,0 +1,64 @@
+          });
+        }
+
+        const hashedPassword = await generateHashPassword(user.password);
```
I see why you did not pass in the object directly into the `createNewUser()` right now, because of the hashedPassword.

PHANTOMS0308 commented on __test__/lib/validation.test.ts:
If you decided to sync your validation logic with Zhiyuan's, then don't forget to add a few test cases to cover those new validation logics

zhiyuan-Z commented on pages/api/user-signup.ts:
```diff
@@ -0,0 +1,64 @@
+          hashedPassword
+        );
+
+        return res.status(201).json({ success: true, data: dbRes });
```
According to the [rfc standards](https://datatracker.ietf.org/doc/html/rfc7231#section-6.3.1), a 200 response would return the object created, a 201 response would return a link or reference (such as an id) to the object created. It seems here the object created is returned. You can decide what data you want to return. Also, it seems that the hashed password is also returned in the data property, which seems unnecessary.

chenyou-H replied:
My bad, because I to make sure the data I created in the database is fine. I am not sure if I should change this, because the person who is doing the integration may need to take a look at the return data.

zhiyuan-Z commented on tsconfig.json:
```diff
@@ -1,22 +1,35 @@
     "esModuleInterop": true,
     "module": "esnext",
-    "moduleResolution": "bundler",
+    "moduleResolution": "node",
```
Can you explain this change?

chenyou-H replied:
this is wired? I never modify this file. Anyway, I will put the 'bundler' value back.

zhiyuan-Z reviewed (commented):
Overall looks good, just slightly modify the response data of the success response and it's good to go.

PHANTOMS0308 reviewed (changes requested):
This pull request should be a reference for future backend development. Separation of Concern was perfectly implemented. However, there were some places that I believe have to be fixed before we can merge.

1. Since in #8 Zhiyuan used camelCase, you should be using camelCase for those fields from the `req.body`
2. In your handler, when database malfunctioned, your handler did not respond
3. There is no data sanitization logics to `firstName` and `lastName`. Both have to be capitalized and lowercased (eg. Shengfeng Wu). This is a business logic, and should not be handled by front end developer in view layer (in React)
4. As for those changes in configuration for API test with Jest, I will make a pull request to present a workable solution for both `js-dom` and `node` Jest testing environment.

PHANTOMS0308 commented:
Another change I would like you to make is that, it is better to place all your util functions inside a folder, and import and export them inside a `index.ts` inside that folder. Because I can foresee there is gonna be many files inside the `/lib`, with a more focused folder structure make code readability higher

For example, you can have `/lib/user/` and inside this user folder you have `index.ts` and other files

PHANTOMS0308 commented on models/User.js:
Also make sure all files are typescript except for root level cofigurations

NobodyLand committed:
Merge https://github.com/VijayIyer/skool-clone into user-api

NobodyLand committed:
modify CRUD mehtod name to make it more self-splanatory

NobodyLand committed:
Change user schema key names fo follow camelCase

NobodyLand committed:
Change some variable names to camelCase
remove json.parse in the api, because next js router can handle it when clinet set their content type to json

NobodyLand committed:
comment out login helper functions

NobodyLand committed:
rm login helper functions bc login api service is completed by other developer

NobodyLand committed:
should handle error for client when client is unable to connect with databse

NobodyLand committed:
Add more rules for validation in order to match frontend validation

NobodyLand committed:
Place all utl functions in a folder

NobodyLand committed:
modiy my test case based on new test rules, environment and so on

NobodyLand committed:
Only return object id when successfully creare an object

NobodyLand committed:
caplize user names first letter

NobodyLand committed:
centralized response logic

NobodyLand committed:
Add a test formatter test

NobodyLand committed:
Rm type test

NobodyLand committed:
Rm some comments

NobodyLand committed:
Some mege conflict due to prettier

NobodyLand committed:
Change prettier setting to avoid future merge conflict

PHANTOMS0308 commented on package.json:
```diff
@@ -8,7 +8,9 @@
     "lint": "next lint",
     "test": "jest",
-    "test:watch": "jest --watch"
+    "test:watch": "jest --watch",
```
this one will not be working though. But you could leave it there. No need to change

PHANTOMS0308 commented on package.json:
```diff
@@ -17,19 +19,25 @@
     "@types/react": "18.2.21",
     "@types/react-dom": "18.2.7",
+    "bcrypt": "^5.1.1",
+    "dotenv": "^16.3.1",
```
We wont be using dotenv

NobodyLand committed:
Merge remote-tracking branch 'upstream/dev' into user-api

NobodyLand committed:
Install the most update package

PHANTOMS0308 reviewed (approved):
Nice, let's move on to the next ticket!

NobodyLand committed:
Merge remote-tracking branch 'upstream/dev' into user-api
